### PR TITLE
Don't be generic over numbers

### PR DESCRIPTION
**Pull request by thomaseizinger**
We only have a limited number of data types in prometheus. Instead of being generic over them, unroll the abstractions manually and one implementation per data type.

### CODE REVIEW

**Comment by mxinden** [comment]
Thanks for providing the patch Thomas! Very much appreciate you exploring alternative approaches.

I am not convinced this is the best way forward.

This patch reduces complexity by cutting out the the `EncodeCounterValue`, `EncodeGaugeValue` and `EncodeExemplarValue` traits plus their corresponding `*Encoder` types. As a side effect that results in a nice reduction in lines of code.

I don't expect the majority of all library users to implement any of the `Encode*Value` traits, but instead, make use of the existing implementations. I do expect some power-users to implement `EnocdeMetric` on their own type. To me, this patch only reduces complexity of the former, but not the latter.

With that in mind, I don't think the reduced complexity of this patch is worth the additional methods on `MetricEncoder`.

Let's keep this pull request open. I want to experiment with the large refactoring in downstream code which might change my mind.

**Comment by thomaseizinger** [comment]
Up to you! My main concern is the use of type parameters for a problem that doesn't need to be "open to extension". The set of types is fixed within prometheus but the usage of type parameters suggests that it is not and should instead be extended.

`serde` follows the same approach btw. It defines a data model and allows you to operate within that but you can't extend it. That makes sense because a particular data format has to operate within a defined data model if it wants to be compatible with the ecosystem.

The bottom line is that I think we are overly abstract here which creates a fair bit of indirection and complexity.
